### PR TITLE
feat: add diagnose/shards api

### PR DESCRIPTION
**Pull request by chunshao90**
## Rationale
Provides `disgnose/shards` api  to check cluster shards status.

## Detailed Changes
* Refactor some code.
* Add `disgnose/shards` api.
```
curl 0:8080/api/v1/diagnose/defaultCluster/shards
{
    "status": "success",
    "data":
    {
        "shard_not_register":
        {
            "0": "127.0.0.1:8832",
            "1": "127.0.0.1:8832",
            "5": "127.0.0.1:8832",
            "6": "127.0.0.1:8832"
        },
        "shard_not_ready":
        {}
    }
}
```

## Test Plan
Manual test and CI.

### CODE REVIEW

File: server/cluster/metadata/cluster_metadata.go
Comment (ShiKaiWi) [inline server/cluster/metadata/cluster_metadata.go:1]: Only codes' location is moved in this file? Maybe you should point out them if any real code changes.

File: server/service/http/api.go [hunk 1]
```diff
@@ -68,6 +68,17 @@ type API struct {
...
 
+type DiagnoseShardStatus struct {
+	NodeName string `json:"node_name"`
+	Status   string `json:"status"`
+}
+
+type DiagnoseShardResult struct {
+	// shardID -> nodeName
+	ShardNotRegister map[storage.ShardID]string              `json:"shard_not_register"`
+	ShardNotReady    map[storage.ShardID]DiagnoseShardStatus `json:"shard_not_ready"`
```
Comment (ShiKaiWi) [inline server/service/http/api.go:79]:
```suggestion
	UnregisteredShards map[storage.ShardID]string              `json:"unregistered_shards"`
	UnreadyShards    map[storage.ShardID]DiagnoseShardStatus `json:"unready_shards"`
```

File: server/storage/types.go [hunk 2]
```diff
@@ -34,9 +34,11 @@ const (
 const (
 	ShardRoleLeader ShardRole = iota + 1
 	ShardRoleFollower
+)
 
-	ShardStatusPartitionOpen ShardStatus = iota + 1
-	ShardStatusReady
+const (
+	ShardStatusReady ShardStatus = iota + 1
```
Comment (ShiKaiWi) [inline server/storage/types.go:40]: Why it starts from 1? Is it ok to start form zero?

File: server/service/http/api.go [hunk 3]
```diff
@@ -586,6 +598,60 @@ func (a *API) listProcedures(req *http.Request) apiFuncResult {
 	return okResult(infos)
 }
 
+func (a *API) diagnoseShards(req *http.Request) apiFuncResult {
+	ctx := req.Context()
+	clusterName := Param(ctx, "name")
```
Comment (ShiKaiWi) [inline server/service/http/api.go:603]: The way to extract the cluster name seems different from the pr description?

File: server/storage/types.go (hunk 2 above)
Comment (chunshao90) [reply server/storage/types.go:40]: The default value in Golang is 0, starting from 1 there will be no ambiguity.

File: server/service/http/api.go [hunk 4]
```diff
@@ -108,6 +120,7 @@ func (a *API) NewAPIRouter() *Router {
 	router.Get("/debug/pprof/block", a.pprofBlock)
 	router.Get("/debug/pprof/goroutine", a.pprofGoroutine)
 	router.Get("/debug/pprof/threadCreate", a.pprofThreadcreate)
+	router.Get("/debug/diagnose/shards/:"+clusterNameParam, wrap(a.diagnoseShards, true, a.forwardClient))
```
Comment (jiacai2050) [inline server/service/http/api.go:123]: `/debug/diagnose/:cluster/shards` is more suitable, in later we may need to check other status of one cluster, such as tables.

File: server/service/http/api.go [hunk 5]
```diff
@@ -586,6 +599,60 @@ func (a *API) listProcedures(req *http.Request) apiFuncResult {
...
+
+	registerNodes, err := a.clusterManager.ListRegisterNodes(ctx, clusterName)
+	if err != nil {
+		log.Error("get cluster failed", zap.Error(err))
+		return errResult(ErrGetCluster, fmt.Sprintf("clusterName: %s, err: %s", clusterName, err.Error()))
+	}
+
+	expectedShardNodes := c.GetShardNodes()
+	// shardName -> shardInfo
+	registerNodesMap := make(map[string][]metadata.ShardInfo)
```
Comment (jiacai2050) [inline server/service/http/api.go:623]: give it a len `len(registerNodes)`

**Review by jiacai2050 — APPROVED** [review]
LGTM

**Pull request merged**